# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/manager/ssl/SSLServerCertVerification.cpp
# Commit: 16350e08d34f
# Full Hash: 16350e08d34ff06c7f46745c6b160c7a7bda9140
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2019-12-06 04:43:57
# Regressor Bug: 1586855
# File Overlap Count: 6
# Description:
#   bug 1586855 - incorporate CRLite filters into cert_storage r=jcj,kjacobs
#   
#   This patch implements CRLite lookups for TLS server certificate revocation
#   information in telemetry-only mode. It adds a new preference
#   "security.pki.crlite_mode" to control the behavior of this feature. Setting
# ==============================================================================

diff -r 64fd40663930 -r 16350e08d34f security/manager/ssl/SSLServerCertVerification.cpp
--- a/security/manager/ssl/SSLServerCertVerification.cpp	Thu Dec 05 22:23:54 2019 +0000
+++ b/security/manager/ssl/SSLServerCertVerification.cpp	Thu Dec 05 22:41:53 2019 +0000
@@ -1014,7 +1014,8 @@
     KeySizeStatus aKeySizeStatus, SHA1ModeResult aSha1ModeResult,
     const PinningTelemetryInfo& aPinningTelemetryInfo,
     const UniqueCERTCertList& aBuiltCertChain,
-    const CertificateTransparencyInfo& aCertificateTransparencyInfo) {
+    const CertificateTransparencyInfo& aCertificateTransparencyInfo,
+    CRLiteTelemetryInfo crliteTelemetryInfo) {
   uint32_t evStatus = (aCertVerificationResult != Success)
                           ? 0  // 0 = Failure
                           : (aEvOidPolicy == SEC_OID_UNKNOWN) ? 1   // 1 = DV
@@ -1053,6 +1054,38 @@
         aBuiltCertChain,
         /*isEV*/ aEvOidPolicy != SEC_OID_UNKNOWN, aCertificateTransparencyInfo);
   }
+
+  switch (crliteTelemetryInfo) {
+    case CRLiteTelemetryInfo::FilterNotAvailable:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::FilterNotAvailable);
+      break;
+    case CRLiteTelemetryInfo::IssuerNotEnrolled:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::IssuerNotEnrolled);
+      break;
+    case CRLiteTelemetryInfo::CertificateTooNew:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::CertificateTooNew);
+      break;
+    case CRLiteTelemetryInfo::CertificateValid:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::CertificateValid);
+      break;
+    case CRLiteTelemetryInfo::CertificateRevoked:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::CertificateRevoked);
+      break;
+    case CRLiteTelemetryInfo::LibraryFailure:
+      Telemetry::AccumulateCategorical(
+          Telemetry::LABELS_CRLITE_RESULT::LibraryFailure);
+      break;
+    case CRLiteTelemetryInfo::NeverChecked:
+      break;
+    default:
+      MOZ_ASSERT_UNREACHABLE("Unhandled CRLiteTelemetryInfo value?");
+      break;
+  }
 }
 
 static void AuthCertificateSetResults(
@@ -1116,6 +1149,7 @@
   SHA1ModeResult sha1ModeResult = SHA1ModeResult::NeverChecked;
   PinningTelemetryInfo pinningTelemetryInfo;
   CertificateTransparencyInfo certificateTransparencyInfo;
+  CRLiteTelemetryInfo crliteTelemetryInfo = CRLiteTelemetryInfo::NeverChecked;
 
   nsTArray<nsTArray<uint8_t>> peerCertsBytes;
   for (CERTCertListNode* n = CERT_LIST_HEAD(peerCertChain);
@@ -1134,11 +1168,12 @@
       certVerifierFlags, Some(peerCertsBytes), stapledOCSPResponse,
       sctsFromTLSExtension, dcInfo, infoObject->GetOriginAttributes(),
       saveIntermediates, &evOidPolicy, &ocspStaplingStatus, &keySizeStatus,
-      &sha1ModeResult, &pinningTelemetryInfo, &certificateTransparencyInfo);
+      &sha1ModeResult, &pinningTelemetryInfo, &certificateTransparencyInfo,
+      &crliteTelemetryInfo);
 
   CollectCertTelemetry(rv, evOidPolicy, ocspStaplingStatus, keySizeStatus,
                        sha1ModeResult, pinningTelemetryInfo, builtCertChain,
-                       certificateTransparencyInfo);
+                       certificateTransparencyInfo, crliteTelemetryInfo);
 
   AuthCertificateSetResults(infoObject, cert, builtCertChain, peerCertChain,
                             certificateTransparencyInfo, evOidPolicy,